docs/source/using_websockets.rst

```diff
@@ -4,30 +4,186 @@ Using WebSockets in Dallinger Experiments
-For example
+It's possible for an experiment class to subscribe to messages on
+additional WebSocket channels. To avoid duplicate subscriptions it's generally
+best to create such subscriptions in your Experiment class's
+:func:`~dallinger.experiment.Experiment.on_launch` method which is only run at
+experiment launch time, or using the experiment's
+:func:`~dallinger.experiment.Experiment.background_tasks`. For example
 
 ::
 
     def on_launch(self):
         from dallinger.experiment_server.sockets import chat_backend
         chat_backend.subscribe(self, 'my_secondary_channel')
+
+An experiment can create and subscribe to channels after launch, but would need
+to be careful to ensure each channel is only ever subscribed once per experiment
+run. This is likely to be difficult because the experiment potentially has many
+instances running concurrently across multiple processes and servers.
+
+Websocket messages are strings consisting of a channel name followed by a `:`
+and then a message payload. The message payload is usually a string representing
+a JSON object. When an experiment receives a message where the payload is a JSON
+object containing either a `node_id` property or a `participant_id` value (the
+properties `participant_id`, `sender`, and `client.participant_id` are treated
+as equivalent), then the experiment will queue such messages to be processed
+asynchronously. Otherwise, the messages will be processed synchronously by an
+experiment instance. In either case, the
+:func:`~dallinger.experiment.Experiment.receive_message` method will handle the
+message.
+
+Client Implementation
+---------------------
+
+The default experiment layout includes a `basic websocket communication library
+<https://www.npmjs.com/package/reconnecting-websocket>`_ which implements a
+`ReconnectingWebSocket` object that can be used to establish channel
+subscriptions, send messages to various channels, and receive messages on
+subscribed channels.
+
+Typically experiments set up a WebSocket connection after completing the initial
+call to `createAgent` using code similar to this
+
+::
+    var broadcast_socket;
+
+    var open_socket = function (channel_id) {
+        var ws_scheme = (window.location.protocol === "https:") ? 'wss://' : 'ws://';
+        // Setup a websocket connection to the channel, passing our worker_id and participant_id
+        socket = new ReconnectingWebSocket(
+            ws_scheme + location.host + "/chat?channel=" + channel_id +"&worker_id=" + dallinger.identity.workerId + '&participant_id=' + dallinger.identity.participantId
+        );
+        // Once the connection is established, send an initial message to the channel
+        socket.onopen(function () {
+            socket.send(channel_id + ':{"message": "Hello world!"}');
+        });
+        // Handle any incoming messages
+        socket.onmessage = function (msg) {
+            // Ignore messages not from the channel subscribed channel
```

Comment:
I think a frequent point of confusion is: what does a channel accomplish in terms of opting in or out of messages received. If I subscribed to "my-channel", why would I still need to perform this check for messages on other channels and ignore them?